### PR TITLE
Tauri

### PR DESCRIPTION
**Pull request by martpie**
Tauri

This is a half-working copy of Museeks ported to Tauri 2.0.

This is highly experimental, many things are broken. Lots of non-idiomatic Rust (and React actually).

```bash
yarn
cargo test --manifest-path src-tauri/Cargo.toml
yarn tauri dev
```

### CODE REVIEW

**Comment by martpie** [comment]
Let me keep track of what's working and what's yet to be fixed. Non-exhaustive.

- [x] Database setup
- [x] Config setup
- [x] Track import
- [x] Loading UI
- [x] Header
  - [x] Controls
  - [x] Search
- [x] Library
  - [x] Playback
  - [ ] Context Menu
  - [ ] Delete tracks
- [ ] Playlists
  - [ ] Auto-import m3u
  - [x] CRUD
    - [x] Create
    - [x] Edit
    - [x] Delete
  - [ ] Playback
- [ ] Settings
  - [ ] Theme selection
  - [ ] `__system`  theme
- [ ] Integrations
  - [ ] Single instance
  - [ ] macOS close without quitting the app
  - [x] Devtools
  - [ ] Dialogs
  - [x] Menu
    - [ ] Menu Events
  - [ ] Native Theme
  - [x] Sleep Blocker
  - [ ] Thumbbar
  - [ ] Window Position

File: src-tauri/src/plugins/cover.rs [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+    let primary_tag = match tagged_file.primary_tag() {
+        Some(tag) => tag,
+        None => return None,
+    };
+
+    let cover = match primary_tag.pictures().first() {
+        Some(cover) => cover,
+        None => return None,
+    };
```
Comment (Serial-ATA) [inline src-tauri/src/plugins/cover.rs:30]: I'd recommend doing `primary_tag.get_picture_type(PictureType::FrontCover)` before falling back to whatever the first picture happens to be.

File: src-tauri/src/plugins/database.rs [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+                        album: album.as_deref().unwrap_or("Unknown").to_string(),
+                        // TODO: get multiple artists/genres: https://github.com/Serial-ATA/lofty-rs/issues/55
+                        artists: vec![artist.as_deref().unwrap_or("Unknown Artist")]
+                            .iter()
+                            .map(|&s| s.into())
+                            .collect(),
+                        genres: vec![genre.as_deref().unwrap_or("")]
+                            .iter()
+                            .map(|&s| s.into())
+                            .collect(),
```
Comment (Serial-ATA) [inline src-tauri/src/plugins/database.rs:388]:
```suggestion
                        artists: tag.get_strings(&ItemKey::TrackArtist).map(ToString::to_string).collect(),
                        genres: tag.get_strings(&ItemKey::Genre).map(ToString::to_string).collect(),
```

**Comment by uklotzde** [comment]
Hi. Your questions regarding Lofty have caught my attention ;) This project looks interesting.

Over the years I have a built a sophisticated backend in Rust, but don't have an appropriate frontend yet. Background: I am a backend developer with almost zero UI skills and this was my starter project for learning Rust back in 2018.

For demonstration purposes and for conveniently synchronizing my library I recently added a [demo app](https://github.com/aoide-org/aoide-rs?tab=readme-ov-file#quickstart) (native, egui, embedded SQLite) with a very basic query parser for searching. Should work on any platform without external dependencies. Does not touch your precious files! Writing back metadata to files is implemented, but not available in the API layer.

For personal use I maintain a read-only integration for Mixxx with a local, standalone web service that is also part of the (bloated) monorepo. Mixxx is my frontend for searching, selecting, and playing music. The (undisclosed) C++/Qt/JSON integration in Mixxx is cumbersome and brittle, not recommended.

Running this branch unfortunately didn't work. It also requires to disable macOS-specific code on Linux, otherwise even the build fails.

Once you have the basics working I am curious to see if it would be possible to attach my backend components. Disclaimer: I have no experience with Tauri.

**Comment by martpie** [comment]
Hi 

> Running this branch unfortunately didn't work. It also requires to disable macOS-specific code on Linux, otherwise even the build fails.

I am mainly developing this port on macOS, and sometimes testing things on Windows on my desktop. I still have to spinup an ubuntu virtual machine and try to see what's failing there. 

Or I should try to fix the CI, which should build the app by itself on win+mac+linux.

I'll add point in the todo-list for those.

**Comment by uklotzde** [comment]
I only get this far on Linux:

image

image

**Comment by Serial-ATA** [comment]
@uklotzde Last year, MediaSession support was disabled by default in `webkit2gtk`: https://github.com/WebKit/WebKit/commit/9a224f3ed860757adc2f8ee13423b4dd45c78737

This is the issue: https://bugs.webkit.org/show_bug.cgi?id=247527

So, you'll have to build `webkit2gtk` with the experimental `ENABLE_MEDIA_SESSION` feature. This is only an issue on Linux.

**Comment by uklotzde** [comment]
> So, you'll have to build `webkit2gtk` with the experimental `ENABLE_MEDIA_SESSION` feature. This is only an issue on Linux.

Yikes. Even if this might be a cumbersome but feasible option during development. It is not an option for any regular user who wants to use the app.

**Comment by martpie** [comment]
Let me push a commit to enable graceful degradation for these features. 

If MPRIS cannot work in Linux, then too bad, but it should not have to hard crash the app.

**Comment by martpie** [comment]
Latest version of this branch should not crash if `mediaSession` is not available

**Comment by martpie** [comment]
Let's merge to a dedicated `tauri` branch :)

**Pull request merged**